File: ua-agent/frontend/src/shared/contracts/library.ts
import { z } from "zod";

import { MaterialEntry, Platform } from "./capture";
import { ErrorEnvelope, SCHEMA_VERSION } from "./error";

export { MaterialEntry };

export const LibraryListFilters = z.object({
  from: z.union([z.string().datetime({ offset: false }), z.null()]),
  to: z.union([z.string().datetime({ offset: false }), z.null()]),
  author: z.union([z.string().max(256), z.null()]),
  platform: z.union([Platform, z.null()]),
  limit: z.number().int().min(1).max(200),
  offset: z.number().int().nonnegative(),
});
export type LibraryListFilters = z.infer<typeof LibraryListFilters>;

export const LibraryListSuccess = z.object({
  schema_version: z.literal(SCHEMA_VERSION),
  ok: z.literal(true),
  entries: z.array(MaterialEntry),
  total: z.number().int().nonnegative(),
  library_total: z.number().int().nonnegative(),
  applied_filters: LibraryListFilters,
});
export type LibraryListSuccess = z.infer<typeof LibraryListSuccess>;

export const LibraryListResult = z.discriminatedUnion("ok", [
  LibraryListSuccess,
  ErrorEnvelope,
]);
export type LibraryListResult = z.infer<typeof LibraryListResult>;

export const LibraryDeleteSuccess = z.object({
  schema_version: z.literal(SCHEMA_VERSION),
  ok: z.literal(true),
  deleted_post_id: z.string().min(1).max(128),
  restored: z.literal(false),
});
export type LibraryDeleteSuccess = z.infer<typeof LibraryDeleteSuccess>;

export const LibraryDeleteResult = z.discriminatedUnion("ok", [
  LibraryDeleteSuccess,
  ErrorEnvelope,
]);
export type LibraryDeleteResult = z.infer<typeof LibraryDeleteResult>;

File: ua-agent/backend/src/ua_agent/contracts/library.py
from __future__ import annotations

from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field

from .capture import MaterialEntry
from .error import ErrorEnvelope


class LibraryListFilters(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True, populate_by_name=True)

    from_: Annotated[str | None, Field(alias="from")]
    to: str | None
    author: Annotated[str | None, Field(max_length=256)]
    limit: Annotated[int, Field(ge=1, le=200)]
    offset: Annotated[int, Field(ge=0)]


class LibraryListSuccess(BaseModel):
    """Success payload of `library list --json`."""

    model_config = ConfigDict(extra="forbid", frozen=True, populate_by_name=True)

    schema_version: Literal["1"] = "1"
    ok: Literal[True] = True
    entries: list[MaterialEntry]
    total: Annotated[int, Field(ge=0)]
    library_total: Annotated[int, Field(ge=0)]
    applied_filters: LibraryListFilters


LibraryListResult = Annotated[
    LibraryListSuccess | ErrorEnvelope,
    Field(discriminator="ok"),
]


class LibraryDeleteSuccess(BaseModel):
    """Success payload of `library delete --json`."""

    model_config = ConfigDict(extra="forbid", frozen=True)

    schema_version: Literal["1"] = "1"
    ok: Literal[True] = True
    deleted_post_id: Annotated[str, Field(min_length=1, max_length=128)]
    restored: Literal[False] = False


LibraryDeleteResult = Annotated[
    LibraryDeleteSuccess | ErrorEnvelope,
    Field(discriminator="ok"),
]
